bson/primitive/primitive.go

```diff
@@ -194,6 +194,24 @@ func (d D) Map() M {
 	return m
 }
 
+// RecursiveMap creates a map from the elements of the D, recursively creating
+// maps from embedded documents and arrays within D.
+func (d D) RecursiveMap() M {
+	m := make(M, len(d))
+	for _, e := range d {
+		key := e.Key
+		switch e := e.Value.(type) {
+		case D:
+			m[key] = e.RecursiveMap()
+		case A:
+			m[key] = e.RecursiveMap()
+		default:
```

Comment:
Do we care about handling other collection types that might have nested `D` types?

E.g.
```go
d := bson.D{
	{"a", bson.D{{"nested", "value"}}},
	{"b", bson.A{bson.D{{"nested", "value"}}}},
	{"c", mongo.Pipeline{bson.D{{"nested", "value"}}}},
	{"d", bson.M{"key": bson.D{{"nested", "value"}}}},
	{"e", []interface{}{bson.D{{"nested", "value"}}}},
	{"f", map[string]interface{}{"key": bson.D{{"nested", "value"}}}},
}
d.RecursiveMap()
```

The nested documents in `"a"` and `"b"` will be converted to `M`, but the rest will not.